src/workerd/api/basics.c++

```diff
@@ -402,10 +409,8 @@ jsg::Ref<AbortSignal> AbortSignal::timeout(jsg::Lock& js, double delay) {
   // completes, whichever comes first.
 
   global.setTimeoutInternal([signal = signal.addRef()](jsg::Lock& js) mutable {
-    auto exception = js.exceptionToJsValue(JSG_KJ_EXCEPTION(FAILED,
+    signal->triggerAbort(js, JSG_KJ_EXCEPTION(DISCONNECTED,
```

Comment:
Does the problem still exist for other call sites of `triggerAbort()` which may not provide a DISCONNECTED exception?

Reply:
But does that mean that the places that use a user-provided error might falsely log it as an uncaught exception?

Reply:
To state my real concern: Changing this to DISCONNECTED feels like it's treating the symptom, taking advantage of logging code that ignores disconnected exceptions. But it seems like the real bug is that this exception is leaking to multiple propagation paths where one of them catches it but the other one deems it uncaught... why is it taking two paths in the first place? I don't quite understand that.

Reply:
Yep, the reason for that is in `fetchImplNoOutputLock`. When there is a body in the fetch request we create a pumpTo promise that is registered as a waitIUntil task in the request context and *separately* return a promise for the `Response` object. One is handling the transmission of the payload to the server, the other is waiting on the response to be ready. Both of these separate promise paths need to be cancelable via the signal. The try/catch in the user code is currently only handling the promise for the response and not the inner waitUntil promise for the outgoing payload. It's that waitUntil promise that is reporting the unhandled exception here when the type is not `DISCONNECTED`.

Reply:
Ah hah! Interesting. I would argue that we should also ignore the exception caught there if it came from the AbortSignal. We really only intend to report exceptions that came from the pump itself.

Reply:
Yeah, that's why the default abort exception (the one that is used if user code does not explicitly provide a reason) uses `DISCONNECTED`. The challenge is with user provided reasons/exceptions. Should those be reported or not? I can definitely see a case where we'd want them to be reported. Should we just *always* make it so that any kj::Exception reported through the `AbortSignal` is `DISCONNECTED`, even if it is converted from a user-provided JS exception/value?

Reply:
No I'm saying I think this code which you screenshotted should ignore the exception if it came from AbortSignal -- even if it's a user-provided exception. The point here is, there's no actual way for the application to catch the exception from the pump, so it's kind of unfair to complain at them that they failed to do something they can't do. Moreover, the exact same exception will also be thrown from `fetch()`, where the application _can_ catch it. So the pump code should assume it's up to the app to catch in there, and so ignore it here.

The more ambiguous question is what to do with exceptions from the pump itself, i.e. that did _not_ come from `AbortSignal`. The application still can't actually catch them, so arguably it's still unfair to complain that it's uncaught. But we don't like to throw away exceptions silently. I'm not really sure what kind of exception could happen here, other than disconnected exceptions, so I'd say we punt on that. But I could be convinced that we should ignore all exceptions here... or maybe that we should ignore all JSG exceptions?

Reply:
Thinking about it, I'd almost prefer if errors that occur in the pumpTo promise there (JSG or otherwise) should actually cause the rest of the request to fail (if they are not `DISCONNECTED`). So, for instance, if the `Response` promise is still pending, if the pumpTo promise fails (for any reason, not just Abort) the `Response` promise would fail also. If the `Response` promise has already resolved, then consuming the `Response` should fail... obviously that's quite a bit more difficult to achieve tho.

The other option would be to forward these into `unhandledrejection` events if the `IoContext` is still alive. That would give user code the opportunity to at least to attach some kind of handling.

Ignoring all JSG exceptions that occur here may be dangerous. For instance, if the body is served by a JavaScript-backed stream that errors during the pump I don't think there's currently any other way really for us to know, if I'm remembering correctly.

For now, let's go ahead with this change and I'll stew a bit on how we can handle these errors more gracefully beyond dropping them on the floor.

Reply:
> Thinking about it, I'd almost prefer if errors that occur in the pumpTo promise there (JSG or otherwise) should actually cause the rest of the request to fail (if they are not DISCONNECTED).

This is tricky because some servers will return an HTTP error and close the connection without reading the entire request. In these cases, we'd much prefer to return the HTTP error to the app, rather than throw an exception from fetch(). If the two things happen at the same time -- an HTTP response is returned, and the upload fails -- we need some way to make sure the HTTP response "wins the race".

That said, if the request upload fails and the response promise stays hanging, I agree it would be best to propagate the request error over to the response rather than continue to hang. This is pretty annoying to get right, though.

> Ignoring all JSG exceptions that occur here may be dangerous... For now, let's go ahead with this change and I'll stew a bit on how we can handle these errors more gracefully beyond dropping them on the floor.

I think that "ignore the error if it came from the AbortSignal" is still clearly correct, even if it was an application error. Nobody wants the exception to be logged if an AbortSignal was used to cancel the request explicitly.

You make a good case that ignoring JSG errors in general may be going too far, but I think ignoring errors from AbortSignal has no down side.

Reply:
> I think that "ignore the error if it came from the AbortSignal" is still clearly correct, even if it was an application error. Nobody wants the exception to be logged if an AbortSignal was used to cancel the request explicitly.

I'm not so sure. If someone is just using the default abort reason, then yes, ignoring it is fine. If someone is taking the time to provide their own `reason` (e.g. `abortController.abort(new Error('boom'))`, then I think it makes sense to at least log it so that use of the custom reason is observable. Just off the top of my head I've seen folks use custom reasons for debugging purposes if there are multiple branches that could lead to cancelation. Or, perhaps, they are using `AbortSignal.any(....)` and they would like to know which if the signals was actually used (by specifying different reasons for each). I don't think we can generalize here that it's always safe to ignore them.

Reply:
If someone uses an AbortSignal to abort a fetch, the call to `fetch()` itself will propagate the exception they gave. There's no need for the pump loop to _also_ log it, because it's already observable through `fetch()`. So again, I think we can just ignore the exception in the pump loop, if it came via AbortSignal.

Reply:
OK, so how can someone use AbortSignal to abort a fetch with an explicit reason but _without_ causing an uncaught exception to be reported? I think people should be able to do that, right?

Reply:
If the user wants to abort a fetch *without* the error being reported they would simply *not* provide any reason to the `abortController.abort()` call. That would use the default error which is marked explicitly as `DISCONNECTED`. I would think that if the user wants to use an explicit reason they would want that to be reported somewhere, otherwise there really isn't a point to using an explicit reason.

Reply:
I actually don't think there's a good argument for treating disconnected errors differently from other errors here in the first place. Usually, in our _internal_ error logging (i.e. to Sentry), we ignore "disconnected" errors because they are common and not necessarily indicative of a problem, but it's kind of a sloppy approach to combating error noise. But in this case, we're not talking about internal-facing logging, we're talking about reporting uncaught exceptions to the app. I don't think it makes sense to treat "disconnected" differently for this purpose -- if the app does in fact want exceptions from the pump loop to be reported as uncaught exceptions at all, surely this includes disconnected exceptions. I think the code here ended up ignoring disconnected errors just because it was following the pattern from internal logging, but it isn't actually the right answer.

But the problem is there's currently no way for the app to actually catch errors from the pump loop, so any error at all is effectively an uncaught exception.

In the common case that the full request is uploaded before a response is received, then we have a clear right answer, which is to report the exception from the response promise. Or if the application is using the JS stream API, that also gives us a clear place to report write errors. But in the case of full-duplex streaming (which is rare), when not using JS streams, if the pump fails after the response was already returned, then we seemingly have nowhere to report this.

It seems like there's a missing API to register an error handler for request upload errors, particularly during full-duplex requests. Does the fetch API spec have anything to say about this?

In the absence of any actual API for receiving these errors, I think we do actually just want to discard them. This just has to be a fact of life when using full duplex, I think. Otherwise we will see more complaints about the inability to silence these.

Reply:
> Does the fetch API spec have anything to say about this?

Remember that the fetch API is written in terms of the ReadableStream spec, assuming a JavaScript backed stream... specifically see here: https://fetch.spec.whatwg.org/#abort-fetch : `If request’s body is non-null and is readable, then cancel request’s body with error.`

With the assumption being, then, that the body is a `ReadableStream`. The implementation of the `ReadableStream` determines how `cancel(reason)` is implemented, including whether `reason` is ignored or not. With this model, implementers have the ability to effectively handle the `reason` in some way -- tho it is perfectly reasonable for an implementation to "handle" it by doing nothing. Note, however, that `cancel(reason)` itself returns a promise that might be rejected.

In our case here, we *could* implement an API for handling unhandled exceptions (perhaps similar to the way Node.js does with [`process.on('uncaughtException')`](https://nodejs.org/dist/latest-v20.x/docs/api/process.html#event-uncaughtexception) but that assumes that we're not using deferred proxy at all (which is a fine assumption for an abortable fetch case but may not hold up consistently in other cases). I don't really think this is the direction we should go.

I still believe we actually *should* be logging these types of things... that is, if the fetch is aborted with an explicitly given reason, we *SHOULD* log that reason. Even if the code can't do anything about it, the reason shows up in the logs. I consider that to be more acceptable than silently dropping the reason on the floor.

Reply:
One idea here: perhaps the issue is not so much that the error is logged, it's that we're logging it specifically as an "uncaught exception" ... which implies the ability to catch it. Perhaps, instead, we should be logging it differently with wording the clearly indicates that the pump was canceled with an explicit reason.

Reply:
> I still believe we actually should be logging these types of things... that is, if the fetch is aborted with an explicitly given reason, we SHOULD log that reason.

Sorry but I just fundamentally do not agree with this.

We have an actual user complaining that they do not want this logging. In their case they are not providing an explicit reason, but I don't see why this makes a difference. An application might be providing an explicit reason specifically so that their own error-handling code can identify intentional cancellations and avoid logging them, in which case we've defeated that by logging it in a way that can't be opted out of.

Reply:
Ok, we'll just have to disagree on that point then. Regardless of that, changing the pump to not to actually report on any error is a different change than what is being done in this PR, which makes sense in either case. The additional modification to the pump error handling can be done in a separate PR.

Reply:
Lots of discussion above 😄 — here is my understanding, please tell me if this is wrong.

*Currently this generates an uncaught exception*

```js
  async function fail() {
	try {
	  return await fetch("https://example.com", {
		method: "POST",
		body: new FormData(),
		signal: AbortSignal.timeout(1),
	  });
	} catch {
	  console.log("caught error");
	}
  }
```

- Uncaught exceptions are never fun to deal with, and we don't have a generic way to handle them in Workers (`window.onunhandledrejection`, `process.onuncaughtexception`, etc.).
- *Instead the example above should generate an exception that is caught by the `catch` handler*
- ^ That's my understanding of what this PR does — is that correct? (If so, seems reasonable to me, and seems to me to address the feedback from https://github.com/cloudflare/workerd/issues/1020#issue-1851787288)

Seems worth adding test cases that demonstrate and exercise what this does?

Reply:
@irvinebroque The exception _is_ actually caught by the catch handler. The problem is that two copies of the exception are thrown. The _other_ copy is not caught by anything, so it is reported as an uncaught exception (which just means it gets logged, which is annoying but not harmful to the application's correctness).

Reply:
I see. And this change makes it such that both exceptions are caught by the `catch` handler? That seems logical to me, and seems like the expectation in https://github.com/cloudflare/workerd/issues/1020#issue-1851787288. Is the concern around breaking changes, and people suddenly seeing a bunch of additional exceptions caught by existing `catch` handlers?

Just seems like unless there's a standards reason, or a scenario where it'd create other problems, avoiding uncaught exceptions are never fun to deal with and should be avoided. No?

Reply:
@irvinebroque ... the change in this PR *just* makes it so that the error/reason passed on when `AbortSignal.timeout()` fires is handled the same way as the default error/reason passed on when the user calls `abortController.abort()` *without* passing an explicit error/reason.

Here's the tl;dr ... When you initiate a fetch with a request body payload, there are two promises created. One of those promises (`P1`) is for the fetch `Response` itself and is resolved when the server sends the response headers. The second promise (`P2`) is for the transmission (pump) of the request body payload to the server. These two promises are resolved independently of each other.

When an `AbortSignal` is used to cancel an operation, both of these promises are arranged to be canceled when the `AbortSignal` is triggered. An `AbortSignal` can be triggered with or without an explicit `reason`. That `reason` is typically an `Error` but can be any JavaScript value. If no `reason` is given explicitly, a default `Error` is created internally that is used as the `reason`. Currently, when we create this default reason we mark it as a `DISCONNECTION` (more on this in a bit).

When we use `AbortSignal.timeout()`, we are creating an `AbortSignal` that is triggered automatically after a given period of time. When it is triggered we create an `Error` explicitly that indicates that the signal was triggered due to a timeout. This PR updates the code so that this timeout `Error` is also marked as a `DISCONNECTION`.

When a fetch operation is using an `AbortSignal`, the promises (`P1` and `P2`) are wrapped so that they can be canceled. Importantly, each is canceled *independently* of the other.  When the `AbortSignal` is triggered, the underlying `reason` is converted into a `kj::Exception` and each of the two promises -- if they haven't already finished -- is cancelled by passing that `kj::Exception` along to reject them.

In the case of `P2` (the promise representing the transmission of our request payload to the server), there is already handling in place so that if the promise is canceled, if the `kj::Exception` is marked as a `DISCONNECTION`, the rejection is ignored. Any other type of `kj::Exception` will be reported as an unhandled exception. So with the change in this PR, if the fetch is canceled using *either* `abortController.abort()` (with no explicit `reason` given) or `AbortSignal.timeout(...)` the rejection of `P2` will be handled silently with no unhandled exception reported -- which *specifically addresses* the bug that was reported in #1020.

The additional concern that is being discussed here, however, is that if a user calls `abortController.abort(reason)` (with `reason` being any JavaScript value) an unhandled exception will be reported when `P2` is rejected *because* the `kj::Exception` that is used to cancel `P2` is not explicitly marked as a `DISCONNECTION`.

@kentonv's argument is that *all* rejections of `P2` that are caused by an `AbortSignal` should either be ignored and not reported at all OR we should have some kind of API that allows those to be reported and handled. Such an API can be problematic. The closest existing API we have would be the `unhandledrejection` event but that would be a bit odd because the `P2` promise is not actually a JavaScript promise, it is a KJ promise. The API approach is also a bit problematic because it wouldn't always be possible for the user to *do* anything about the rejection. Currently we log the error, which I contend is useful information for the user, as there may not be any other way for the user to know why the request body payload failed to complete.

This change in the error handling of P2 rejections, however, is beyond the scope of *this* specific PR. All *this* specific PR does is make it so that `AbortSignal.timeout()` is handled the same way as `abortController.abort()`. If we decide that the error handling for P2 rejections needs to be changed, that can (and I'd argue *should*) be done in a separate PR.

Reply:
Another consideration, if we were to eliminate all logging here, then in the case where someone is using `AbortSignal.any(...)` it would become tricky to know exactly which `AbortSignal` instance triggered the actual abort.

```
const signal = AbortSignal.any([AbortSignal.timeout(1000), ac.signal, otherAbortSignal, ...]);
```

Reply:
I'll accept merging this PR but should we add a test that verifies the caught exception has the right description? It looks like we do not have such a test today in either the internal or external codebase. I vaguely worry about the fact that DISCONNECTED errors usually get the generic description "Network connection lost."

Reply:
We currently have no way of testing the out-of-band log output in workerd tests but I will open a separate internal PR that adds a test specifically for the log output in these cases.

Reply:
You can catch the exception thrown by fetch().

Reply:
We already have tests verifying the exception thrown by fetch.